bon-macros/src/builder/builder_gen/start_fn.rs

```diff
@@ -96,8 +96,31 @@ impl super::BuilderGenCtx {
             };
 
         let const_ = &self.const_;
+        // add the `clippy::needless_lifetimes` lint if before rust version 1.87
+        // Rust version 1.87 includes a clippy change where `needless_lifetimes`
+        // was split with the more complex part of the lint going to
+        // `elidable_lifetime_names`. For versions since 1.87 we want to block
+        // `elidable_lifetime_names` (See
+        // https://github.com/elastio/bon/pull/341#discussion_r2398893516 for
+        // an explanation).
+        #[rustversion::before(1.87)]
+        fn get_needless_lifetime_lint_annotation() -> TokenStream {
+            quote! {
+                #[allow(clippy::needless_lifetimes)]
+            }
+        }
+        #[rustversion::since(1.87)]
+        fn get_needless_lifetime_lint_annotation() -> TokenStream {
+            quote! {
+                #[allow(clippy::elidable_lifetime_names)]
+            }
+        }
```

Comment:
Could be simplified slightly to avoid some repetition like this:

```rust
        let needless_lifetime_lint = if rustversion::cfg!(before(1.87)) {
            format_ident!("needless_lifetimes")
        } else {
            format_ident("elidable_lifetime_names")
        };
        
        // inside final syn::parse_quote_spanned
        #[allow(clippy::#needless_lifetime_lint)]
```